### ISSUE TITLE
Looking to call the `extender` from within the `creator`

### ISSUE THREAD

**Issue by cormacpayne** [buildpacks/lifecycle#1155]
### Description

The `creator` doesn't appear to have support for calling the `extender`; digging around a little, it appears this behavior is intentional and prevented in a couple of ways:
- [Instantiating the `creator`'s directory store with an empty extender directory](https://github.com/buildpacks/lifecycle/blob/main/cmd/lifecycle/creator.go#L110)
- [Raising an error in `pack` if the `creator` should be used when running the lifecycle with extensions](https://github.com/buildpacks/pack/blob/main/internal/build/lifecycle_execution.go#L285-L287)

In our particular case, we're looking to call the `creator` from within a container based on our builder (which is bundled with extensions) to build and push an image to a registry.

### Proposed solution

Optimistically, the `creator` would include a flag (or flow) to check for extensions within the builder and would swap the call to the `builder` for a call to the `extender`. In the alternatives section below, I outlined our behavior for calling `extender` directly but continuing to call the `creator` would be optimal for us.

I'm not entirely sure if giving the `extender` from within the `creator` is feasible from a permissions point of view, but would love to hear additional thoughts on the subject and ways that this could be solved, if it was an issue that was looking to be solved in the future when extensions were introduced.

### Describe alternatives you've considered

Since the `extender` cannot be called from the `creator` (as outlined above), we've tried to call the individual phase commands of the `creator` command, which involves calling the `extender` rather than the `builder` to carry out our desired flow. When calling the `extender` directly, it's able to get to the point where the build extension is used, but fails due to an "Operation not permitted" error when attempting to run the first line of our extension's `Dockerfile` snippet, as seen below:

<img width="1248" alt="image (2)" src="https://github.com/buildpacks/lifecycle/assets/4462409/5402a573-72c0-4a25-a6d2-d72d15265aae">

My initial hunch is that this is due to the `extender` not being called from within an individual container with root access, as is done by [`pack` when it calls the `extender` during `build`](https://github.com/buildpacks/pack/blob/main/internal/build/lifecycle_execution.go#L684), but would love any additional insight. In our case, we're currently not able to spin up a container from within this container hosting the builder image where we're calling the individual `creator` commands, so this isn't seemingly a route that we'll be able to pursue further.

### Additional context

I'm happy to share any additional information or context that may be needed!

**Comment by natalieparellano** [comment]
@cormacpayne re `Operation not permitted` - are you invoking the extender as root?

**Comment by natalieparellano** [comment]
Regarding running the `extender` from the `creator` - this could probably be done in limited circumstances, but requires some care. Some thoughts are below - 

The easiest use case to support would be image extensions that simply switch the runtime base image to a new base. This actually doesn't require running the extender at all (it is done in the detector) but would involve modifying some validations that error out if extensions are provided and the creator is invoked. 

The next use case would be image extensions that extend the build-time base image. This is completely doable but we would likely need to retain root privileges through analyze-detect-restore-extend(build), which would involve removing validation in the detector that refuses to run as root, which could be somewhat dangerous. We would need to add some code to ensure that buildpack (./bin/detect) processes themselves are run as the CNB user even when the lifecycle is root. But it would still mean that some lifecycle activities would happen with elevated permissions compared to what exists today. Another downside would be that it would be impossible to isolate registry credentials from Dockerfile instructions, because those creds would exist in the same container where extension is running (as root). Today we protect creds by running the restorer (which downloads stuff) in a separate container from the extender. This could potentially be mitigated by the platform pre-downloading the needed stuff.

The hardest use case to support would be run image extension. Simply because, the lifecycle doesn't run in the context of the run image for all the other phases. At one point, we tossed out the "crazy idea" of saving off the build image filesystem after the build phase, downloading & extracting the run image filesystem, running run image extension, saving off the extended run image filesystem, extracting the build image filesystem again, and continuing on to the export phase. It still sounds crazy TBH but if there's a real need for it we could explore it.

**Comment by cormacpayne** [comment]
@natalieparellano Thanks for the detailed response as always, Natalie! Please find responses below:

> re `Operation not permitted` - are you invoking the extender as root?

For the information outlined in the original description of this issue, I was invoking the extender as a non-root user -- I did follow this up by creating a builder whose build stack was executed as root and attempted to run the same script (with each phase ran as their corresponding lifecycle command), and that failed during the detector phase when calling the [`Privileges()`](https://github.com/buildpacks/lifecycle/blob/main/cmd/lifecycle/detector.go#L53-L59) function. This _seemingly_ was the only phase (that we're calling, aka everything but builder) that had this type of check, so I'm not sure if any other phase would've run into issues when executing as root, but you have _much_ more context about the codebase, so I'll defer to you here.

> The easiest use case to support would be image extensions that simply switch the runtime base image to a new base.

I'm glad you brought this up because this was something I was curious about when digging into these scenarios a bit; specifically, if we didn't have a build extensions, would this flow work with _just_ run extensions since it shouldn't be performing any steps that required elevated permissions (AFAIK). I didn't have a chance to try it out, but this is seemingly something that would've worked from calling the individual lifecycle commands, but not the creator (as you already mentioned with the validation required).

> The next use case would be image extensions that extend the build-time base image.

This is a lot of good information to know about what's going on security-wise behind-the-scenes; if there was a way to run _only_ the detector as the CNB user as you mentioned, I think the rest of the phases _should_ work for us (as mentioned above). We're also exploring some options on our side that would allow us to spin up another container within the container that's hosting the builder image, and these sub-containers would be responsible for running each phase with different permission levels, similar to how you all are calling phases via the creator. I'll keep you posted on any updates from these findings since this may not require any changes to the creator but rather us just having to call each phase individually.

> The hardest use case to support would be run image extension.

I don't _think_ we need any additional changes from the run image extension -- correct me if I'm wrong, but I think we should be OK with just the basic base image swapping as denoted above from the second block. I'm not sure if the run image extension scenario you're outlining here is in reference to something more than the base image swapping.

**Comment by natalieparellano** [comment]
> This seemingly was the only phase (that we're calling, aka everything but builder) that had this type of check,

You are correct! The builder does have this check, but not the extender (the extender will drop privileges before executing anything related to buildpack ./bin/build execution).

> these sub-containers would be responsible for running each phase with different permission levels

That seems to be the best option security-wise. Using the creator requires making some assumptions of trust, and using the extender even more so.

> we should be OK with just the basic base image swapping as denoted above from the second block

That's great news! The "simply switching the runtime base image" case would be fairly easy to implement. If there's broad support for it, we could even target it for the next release (0.18.0).

**Comment by natalieparellano** [comment]
A small follow-up to this - I recently encountered the dreaded "operation not permitted" error when trying to add extensions support to [kpack](https://github.com/buildpacks-community/kpack/pull/1332), even though I was running the extender as root. It turns out I needed to add `SETGID` and `SETUID` capabilities to the build pod (see [here](https://github.com/buildpacks-community/kpack/pull/1332/files#diff-796b50408de7dbf79786503ffc9edfbc1eb29abc4c36902f37c6c5714606333bR763)) due to the `USER root` instruction in my Dockerfile. Platforms with different Dockerfiles may have different requirements (and we're still digging to determine if these capabilities are truly necessary) but I thought I'd share this info in case it helps with your investigation.

**Comment by c0d1ngm0nk3y** [comment]
@natalieparellano 

We are looking into this and I have some questions, you might be able to answer...

1.)
> The hardest use case to support would be run image extension. Simply because, the lifecycle doesn't run in the context of the run image for all the other phases.

We were surprised to read in the spec that the `extent (run)` must be called with the run environment. Why exactly? All the work of the extension is already done and "only" the `kaniko` call is missing, so the environment should not matter. Or what is the background for this requirement? 

2. )
Doesn't the `creator` have to call `extend` from `PLATFORM_API=0.13` forward (once extensions are not experimental anymore)? Otherwise the `creator` would not be spec-compliant. Right?

**Comment by natalieparellano** [comment]
> "only" the kaniko call is missing, so the environment should not matter

We use a little-known kaniko [option](https://github.com/buildpacks/lifecycle/blob/200b00a4a421d95f6d781d1d0a00fffcc3455771/internal/extend/kaniko/dockerfile_applier.go#L85) to avoid unpacking the initial filesystem (the `FROM <image>`). This only works because we are running in the context of the image being extended, and multi-stage Dockerfiles are not permitted. 

The reason is to avoid exposing registry credentials to Dockerfiles generated from image extensions (see for example https://github.com/GoogleContainerTools/kaniko/issues/2153).

**Comment by natalieparellano** [comment]
> Doesn't the creator have to call extend from PLATFORM_API=0.13 forward (once extensions are not experimental anymore)? Otherwise the creator would not be spec-compliant. Right?

I think the `extender` is still considered optional. But, we should make it clear in the spec that the creator does not call the extender.

**Comment by loewenstein** [comment]
@natalieparellanoit looks a bit like some platforms would need a re-iteration over what you stated in a comment above...
> At one point, we tossed out the "crazy idea" of saving off the build image filesystem after the build phase, downloading & extracting the run image filesystem, running run image extension, saving off the extended run image filesystem, extracting the build image filesystem again, and continuing on to the export phase. It still sounds crazy TBH but if there's a real need for it we could explore it.

My weakest spot in understanding the full scope of constraints for run image extensions is probably which run image needs to be in the context of the run image extender? Or, more precisely, how dynamic this information is by nature. If it is potentially decided by one of the run image extensions - then platforms need to be able to dynamically schedule containers, right? For Kubernetes based platforms, we would be then talking about creating individual Pods for the different phases, correct?

**Comment by natalieparellano** [comment]
> For Kubernetes based platforms, we would be then talking about creating individual Pods for the different phases, correct?

That is probably going to need to happen. https://github.com/buildpacks-community/kpack/pull/1332 is our (somewhat outdated) attempt to get build image extension + run image switching to work in kpack.

**Comment by chkpnt** [comment]
Just for my understanding: Is this issue the reason I'm failing to switch to the UBI stack with Gradle's `bootBuildImage` task?

I switched to
```
builder = paketocommunity/builder-ubi-base:latest
runImage = paketobuildpacks/run-java-11-ubi8-base:latest
```
which led to the following error:

```
 > Running creator
    [creator]     ===> ANALYZING
    [creator]     Image with name "my-app:latest" not found
    [creator]     ===> DETECTING
    [creator]     target distro name/version labels not found, reading /etc/os-release file
    [creator]     9 of 30 buildpacks participating
    [creator]     paketo-community/ubi-java-extension     0.2.0
    [creator]     paketo-community/ubi-java-buildpack     0.1.1
    [creator]     paketo-buildpacks/ca-certificates       3.10.0
    [creator]     paketo-buildpacks/bellsoft-liberica     11.1.0
    [creator]     paketo-buildpacks/syft                  2.9.0
    [creator]     paketo-buildpacks/executable-jar        6.13.0
    [creator]     paketo-buildpacks/dist-zip              5.10.0
    [creator]     paketo-buildpacks/spring-boot           5.33.0
    [creator]     paketo-buildpacks/environment-variables 4.9.0
    [creator]     ERROR: failed to detect: detected order contains extensions which is not supported by the creator
```

**Comment by natalieparellano** [comment]
@chkpnt that's right - you would need to invoke the 5 phases separately in order to use extensions
